MetroidPrimeClient.py

```diff
@@ -0,0 +1,206 @@
+
+
+class MetroidPrimeCommandProcessor(ClientCommandProcessor):
+    def __init__(self, ctx: CommonContext):
+        super().__init__(ctx)
+
+
+class MetroidPrimeContext(CommonContext):
+    command_processor = MetroidPrimeCommandProcessor
+    game_interface: MetroidPrimeInterface
+    game = "Metroid Prime"
+    items_handling = 0b111
+    dolphin_sync_task = None
+
+    def __init__(self, server_address, password):
+        super().__init__(server_address, password)
+        self.game_interface = MetroidPrimeInterface(logger)
+
+    def on_deathlink(self, data: Utils.Dict[str, Utils.Any]) -> None:
+        super().on_deathlink(data)
+        logger.info()
+
+    async def server_auth(self, password_requested: bool = False):
+        if password_requested and not self.password:
+            await super(MetroidPrimeContext, self).server_auth(password_requested)
+        await self.get_username()
+        await self.send_connect()
+
+
+async def dolphin_sync_task(ctx: MetroidPrimeContext):
+    logger.info("Starting Dolphin connector")
+    while not ctx.exit_event.is_set():
+        try:
+            if ctx.game_interface.is_connected() and ctx.game_interface.is_in_playable_state():
+                await _handle_game_ready(ctx)
+            else:
+                await _handle_game_not_ready(ctx)
+        except Exception as e:
+            if isinstance(e, DolphinException):
+                logger.error(str(e))
+            else:
+                logger.error(traceback.format_exc())
+
+            logger.info("Attempting to reconnect to Dolphin")
+            await ctx.disconnect()
+            await asyncio.sleep(3)
+            continue
+
+
+def inventory_item_by_network_id(network_id: int, current_inventory: dict[str, InventoryItemData]) -> InventoryItemData:
+    for item in current_inventory.values():
+        if item.code == network_id:
+            return item
+    return None
+
+
+def get_total_count_of_item_received(network_id: int, items: list[NetworkItem]) -> int:
+    count = 0
+    for network_item in items:
+        if network_item.item == network_id:
+            count += 1
+    return count
+
+
+async def send_checked_locations(ctx: MetroidPrimeContext):
+    pass
+
+
+async def handle_receive_items(ctx: MetroidPrimeContext):
+    current_items = ctx.game_interface.get_current_inventory()
+
+    # Handle Single Item Upgrades
+    for network_item in ctx.items_received:
+        item_data = inventory_item_by_network_id(
+            network_item.item, current_items)
+        if item_data is None:
+            logger.debug(
+                f"Item with network id {network_item.item} not found in inventory. {network_item}")
+            continue
+        if item_data.max_capacity == 1 and item_data.current_amount == 0:
+            logger.debug(f"Giving item {item_data.name} to player")
+            ctx.game_interface.give_item_to_player(item_data.id, 1, 1)
+
+    # Handle Missile Expansions

```

Comment:
I'm not in love with the size of this function, but I figured rather than creating an abstraction it'd be more pragmatic to keep them separate since each of these 3 different powerups have a slightly different effect when you receive them (gain 5 missiles, gain 1 bomb, restore health). Open to suggestions here though for sure

Reply:
This is an area where we can also gate the player from receiving any Missile or PB expansions if they haven't received the Missile Launcher or Power Bomb Main items yet.